packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -762,6 +772,217 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
+      expectedBulkGetResults.map<Promise<ExpectedBulkDeleteResult>>(
+        async (expectedBulkGetResult) => {
+          if (isLeft(expectedBulkGetResult)) {
+            return { ...expectedBulkGetResult };
+          }
+          const { esRequestIndex, id, type, version } = expectedBulkGetResult.value;
+
+          let namespaces;
+          let versionProperties;
+
+          if (esRequestIndex !== undefined) {
+            const indexFound = bulkGetResponse?.statusCode !== 404;
+
+            const actualResult = indexFound
+              ? bulkGetResponse?.body.docs[esRequestIndex]
+              : undefined;
+
+            const docFound = indexFound && isMgetDoc(actualResult) && actualResult.found;
+
+            if (!docFound) {
+              return {
+                tag: 'Left',
+                value: {
+                  id,
+                  type,
+                  error: errorContent(
+                    SavedObjectsErrorHelpers.createGenericNotFoundError(type, id)
+                  ),
+                },
+              };
+            }
+            if (
+              // @ts-expect-error MultiGetHit is incorrectly missing _id, _source
+              !this.rawDocExistsInNamespace(actualResult, namespace) &&
+              !force
+            ) {
+              // the document exists but not in the namespace for this client. Deleting the doc is still possible but one needs to force the action.
+              return {
+                tag: 'Left',
+                value: {
+                  success: false,
+                  id,
+                  type,
+                  error: errorContent(
+                    SavedObjectsErrorHelpers.createBadRequestError(
+                      `Unable to delete saved object id: ${id}, type: ${type} that exists in multiple namespaces, use the "force" option to delete all saved objects`
+                    )
+                  ),
+                },
+              };
+            }
+            // @ts-expect-error MultiGetHit is incorrectly missing _id, _source
+            namespaces = actualResult!._source.namespaces ?? [
+              // @ts-expect-error MultiGetHit is incorrectly missing _id, _source
+              SavedObjectsUtils.namespaceIdToString(actualResult!._source.namespace),
+            ];
+            versionProperties = getExpectedVersionProperties(version);
+          } else {
+            if (this._registry.isSingleNamespace(type)) {
+              namespaces = [SavedObjectsUtils.namespaceIdToString(namespace)];
+            }
+            versionProperties = getExpectedVersionProperties(version);
+          }
+
+          const expectedResult = {
+            type,
+            id,
+            namespaces,
+            esRequestIndex: bulkDeleteRequestIndexCounter++,
+          };
+
+          bulkDeleteParams.push({
+            delete: {
+              _id: this._serializer.generateRawId(namespace, type, id),
+              _index: this.getIndexForType(type),
```

Comment:
NIT: we already generated the _id and _index for the objects, L813

```ts
 .map(({ value: { type, id } }) => ({
        _id: this._serializer.generateRawId(namespace, type, id),
        _index: this.getIndexForType(type),
        _source: ['type', 'namespaces'],
      }));
```

We could eventually try to re-use them if we can store them in a var. If that's too complex given the various loops of the function, nevermind it.

Reply:
I'll extract these as part of the refactor